src/api.py

```diff
@@ -42,59 +64,199 @@ def config_cls(cls) -> Type[Config]:
     config: PromptGenerationPluginConfig
 
     def __init__(
-        self,
-        client: Steamship = None,
-        config: Dict[str, Any] = None,
-        context: InvocationContext = None,
+            self,
+            client: Steamship = None,
+            config: Dict[str, Any] = None,
+            context: InvocationContext = None,
     ):
         super().__init__(client, config, context)
         openai.api_key = self.config.openai_api_key
 
+    def completion_with_retry(self, **kwargs: Any) -> Any:
+        """Use tenacity to retry the completion call."""
+        min_seconds = 4
+        max_seconds = 10
+
+        # Wait 2^x * 1 second between each retry starting with
+        # 4 seconds, then up to 10 seconds, then 10 seconds afterwards
+
+        @retry(
+            reraise=True,
+            stop=stop_after_attempt(self.config.max_retries),
+            wait=wait_exponential(multiplier=1, min=min_seconds, max=max_seconds),
+            retry=(
+                    retry_if_exception_type(openai.error.Timeout)
+                    | retry_if_exception_type(openai.error.APIError)
+                    | retry_if_exception_type(openai.error.APIConnectionError)
+                    | retry_if_exception_type(openai.error.RateLimitError)
+            ),
```

Comment:
Good work on finding a nice library to elegantly deal with this, but man this code is weird looking.